src/bpfilter/xlate/cli.c

```diff
@@ -56,6 +60,206 @@ int _bf_cli_ruleset_flush(const struct bf_request *request,
+    }
+
+    return 0;
+}
+
+static int _bf_cli_get_chain_list(bf_list **chain_list)
+{
+    bf_list_ops ops = {// chain_list will only contain the chains, not own them.
+                       .free = (bf_list_ops_free)bf_nop_free,
+                       .marsh = (bf_list_ops_marsh)bf_chain_marsh};
+
+    int r = bf_list_new(chain_list, &ops);
+    if (r < 0)
+        return bf_err_r(r, "Failed to create the chain list");
+
+    // Loop over codegens which correspond to hooks
+    for (int i = 0; i < _BF_HOOK_MAX; ++i) {
+        struct bf_cgen *cgen;
+
+        cgen = bf_ctx_get_cgen(i, NULL);
+
+        if (!cgen || !cgen->chain)
+            continue;
+
+        r = bf_list_add_tail(*chain_list, cgen->chain);
+        if (r)
+            return bf_err_r(r, "Failed to add chain to list");
+    }
+
+    return 0;
+}
+
+static int _bf_cli_get_counters_marsh(struct bf_marsh **counter_marsh,
+                                      bf_list *chain_list, bool with_counters)
+{
+    int r;
+    size_t num_counters = 0;
+    struct bf_counter *counters = NULL;
+
+    bf_assert(bf_list_size(chain_list) > 0);
+
+    // If we don't want counters, return an empty marsh
+    if (!with_counters) {
+        r = bf_marsh_new(counter_marsh, NULL, 0);
+        return r;
+    }
+
+    /* Each chain has a policy counter and an error counter.
+     * Each rule has a counter, though it may be unused. */
+    num_counters =
+        (2 * bf_list_size(chain_list)) + _bf_cli_num_rules(chain_list);
+
+    // We must have at least have the two chain counters
+    bf_assert(num_counters >= 2);
+
+    counters = calloc(num_counters, sizeof(struct bf_counter));
+    if (!counters)
+        bf_err_r(-ENOMEM, "Failed to allocate memory for counters\n");
+
+    r = _bf_get_ctr_vals(chain_list, counters);
+    if (r < 0)
+        return bf_err_r(r, "Could not get ctr vals\n");
+
+    size_t marsh_size = num_counters * sizeof(struct bf_counter);
+
+    r = bf_marsh_new(counter_marsh, counters, marsh_size);
+    if (r < 0)
+        if (r < 0)
+            return bf_err_r(r, "Failed to make new marsh\n");
+
+    free(counters);
+
+    return 0;
+}
+
+static int _bf_cli_get_rules(const struct bf_request *request,
+                             struct bf_response **response)
+{
+    _cleanup_bf_marsh_ struct bf_marsh *chain_list_marsh = NULL;
+    _cleanup_bf_marsh_ struct bf_marsh *counter_marsh = NULL;
+    _cleanup_bf_marsh_ struct bf_marsh *result_marsh = NULL;
+    int r;
+
+    // Empty context, nothing to return
+    if (bf_ctx_is_empty()) {
+        return bf_response_new_success(response, NULL, 0);
+    }
+
+    {
+        _cleanup_bf_list_ bf_list *chain_list = NULL;
+
+        r = _bf_cli_get_chain_list(&chain_list);
+        if (r < 0)
+            return bf_err_r(r, "Failed to create the chain list");
+
+        // Because the context is not empty, we must have at least one chain
+        bf_assert(bf_list_size(chain_list) > 0);
```

Comment:
Assertions are used to test requirements early in a function. Here, we don't expect `chain_list` to be empty, as we know the context is not empty. If somehow `chain_list` is empty, it means something failed previously, and it would be caught through a function's return value.